packages/itwin/tree-widget/src/tree-widget-react/components/trees/classifications-tree/internal/ClassificationsTreeVisibilityHandler.ts

```diff
@@ -181,6 +206,62 @@ class ClassificationsTreeVisibilityHandlerImpl implements HierarchyVisibilityHan
     });
   }
 
+  private getFilteredNodeVisibility(props: GetFilteredNodeVisibilityProps) {
+    return from(this.getVisibilityChangeTargets(props)).pipe(
+      mergeMap(({ classificationIds, classificationTableIds, elements2d, elements3d }) => {
+        const observables = new Array<Observable<VisibilityStatus>>();
+        if (classificationTableIds?.size) {
+          observables.push(this.getClassificationTableDisplayStatus({ classificationTableIds: [...classificationTableIds] }));
+        }
+
+        if (classificationIds?.size) {
+          observables.push(this.getClassificationDisplayStatus({ classificationIds: [...classificationIds] }));
+        }
+        if (elements2d?.size) {
+          observables.push(
+            from(elements2d).pipe(
+              releaseMainThreadOnItemsCount(50),
+              mergeMap(([elementKey, elementIds]) => {
+                const { modelId, categoryId } = parseElementKey(elementKey);
+                assert(modelId !== undefined);
+                assert(categoryId !== undefined);
```

Comment:
I don't like that we have to do this here, outside of `FilteredTree.ts`, where the format of `ElementKey` is defined (BTW, `ElementKey` for `${ModelId}-${CategoryId}` type is a bad name). Could `FilteredTree.getVisibilityChangeTargets` function take care of this processing and instead return something like the following?

```ts
{
    elements2d?: Array<{ modelId: Id64String; categoryId: Id64String; elementIds: Set<Id64String> }>;
}
```